### PR TITLE
z3: add z3/4.8.7 recipe

### PR DESCRIPTION
**Pull request by madebr**
Specify library name and version:  **z3/4.8.7**

A shared mpir depenency is required because of https://github.com/wbhart/mpir/issues/286

- [x] I've read the [guidelines](https://github.com/conan-io/conan-center-index/wiki#how-to-submit-a-pull-request) for contributing.
- [X] I've followed the [PEP8](https://www.python.org/dev/peps/pep-0008/) style guides for Python code in the recipes.
- [X] I've used the [latest](https://github.com/conan-io/conan/releases/latest) Conan client version.
- [x] I've tried at least one configuration locally with the
      [conan-center hook](https://github.com/conan-io/hooks.git) activated.

### CODE REVIEW

**Comment by madebr** [comment]
@uilianries 
I've added `mpir:shared=True` as a workaround for https://github.com/wbhart/mpir/issues/286, but it appears a shared mpir is not available for all configurations that are run on cci.

**Comment by uilianries** [comment]
@madebr It's because we don't provide shared + runtime MT, because it's a configuration for static library only. You will need to filter it on this recipe using a conditional.

File: recipes/z3/all/conanfile.py [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+
+        self.cpp_info.components["libz3"].libs = ["libz3" if self.settings.os == "Windows" else "z3"]
+        self.cpp_info.components["libz3"].requires = ["mpir::mpir"]
+        # Z3COnfig.cmake generates a z3::libz3 target
+        self.cpp_info.components["libz3"].names["cmake_find_package"] = "libz3"
+        self.cpp_info.components["libz3"].names["cmake_find_package_multi"] = "libz3"
+        if not self.options.shared:
+            if self.settings.os  =="Linux":
+                self.cpp_info.components["libz3"].system_libs.append("pthread")
+        # FIXME: this generates a Z3::libz3 target instead of z3::libz3.
```
Comment (jgsogo) [inline recipes/z3/all/conanfile.py:96]: We can fix this `package_info` using `cpp_info.filename`, it allows us to create different names for the file and the namespace.

File: recipes/z3/all/conanfile.py [hunk 2]
```diff
@@ -0,0 +1,97 @@
...
+    def package_info(self):
+        # z3 generates a Z3Config.cmake file
+        self.cpp_info.names["cmake_find_package"] = "Z3"
+        self.cpp_info.names["cmake_find_package_multi"] = "Z3"
+
+        self.cpp_info.components["libz3"].libs = ["libz3" if self.settings.os == "Windows" else "z3"]
+        self.cpp_info.components["libz3"].requires = ["mpir::mpir"]
+        # Z3COnfig.cmake generates a z3::libz3 target
+        self.cpp_info.components["libz3"].names["cmake_find_package"] = "libz3"
+        self.cpp_info.components["libz3"].names["cmake_find_package_multi"] = "libz3"
```
Comment (madebr) [inline recipes/z3/all/conanfile.py:92]:
```suggestion
        self.cpp_info.components["libz3"].names["cmake_find_package"] = "libz3"
        self.cpp_info.components["libz3"].names["cmake_find_package_multi"] = "libz3"
        self.cpp_info.components["libz3"].filenames["cmake_find_package"] = "Z3"
        self.cpp_info.components["libz3"].filenames["cmake_find_package_multi"] = "Z3"
```

File: recipes/z3/all/conanfile.py [hunk 3]
```diff
@@ -0,0 +1,97 @@
...
+        self.copy("LICENSE.txt", src=self._source_subfolder, dst="licenses")
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        # z3 generates a Z3Config.cmake file
+        self.cpp_info.names["cmake_find_package"] = "Z3"
+        self.cpp_info.names["cmake_find_package_multi"] = "Z3"
```
Comment (madebr) [inline recipes/z3/all/conanfile.py:86]:
```suggestion
        self.cpp_info.names["cmake_find_package"] = "z3"
        self.cpp_info.names["cmake_find_package_multi"] = "z3"
```

File: recipes/z3/all/conanfile.py (hunk 1 above)
Comment (madebr) [reply recipes/z3/all/conanfile.py:96]: Good idea.

I find the way to create cmake imported targets in conan lacking though.
The current way to create a `namespace::target`in `FindName.cmake` is limiting.
```cmake
def package_info(self):
    self.cpp_info.names["cmake_find_package"] = "namespace"
    self.cpp_info.components["comp"].names["cmake_find_package"] = "target"
    self.cpp_info.components["comp"].filenames["cmake_find_package"] = "Name"
```


How can I, **in one `package_info` function call**, create `namespace1::target1` in `FindName1.cmake` AND create `namespace2::target2` in `FindName2.cmake`?

File: recipes/z3/all/conanfile.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+        self.cpp_info.names["cmake_find_package"] = "z3"
+        self.cpp_info.names["cmake_find_package_multi"] = "z3"
+
+        self.cpp_info.components["libz3"].libs = ["libz3" if self.settings.os == "Windows" else "z3"]
+        self.cpp_info.components["libz3"].requires = ["mpir::mpir"]
+        # Z3COnfig.cmake generates a z3::libz3 target
+        self.cpp_info.components["libz3"].names["cmake_find_package"] = "libz3"
+        self.cpp_info.components["libz3"].names["cmake_find_package_multi"] = "libz3"
+        self.cpp_info.components["libz3"].filenames["cmake_find_package"] = "Z3"
+        self.cpp_info.components["libz3"].filenames["cmake_find_package_multi"] = "Z3"
```
Comment (jgsogo) [inline recipes/z3/all/conanfile.py:94]: No effect at this level, `filenames` can only be used at the root level: `cpp_info.filenames`

```suggestion
```

File: recipes/z3/all/conanfile.py [hunk 5]
```diff
@@ -0,0 +1,98 @@
...
+        self.copy("LICENSE.txt", src=self._source_subfolder, dst="licenses")
+        cmake = self._configure_cmake()
+        cmake.install()
+
+        tools.rmdir(os.path.join(self.package_folder, "lib", "cmake"))
+
+    def package_info(self):
+        # z3 generates a Z3Config.cmake file
+        self.cpp_info.names["cmake_find_package"] = "z3"
+        self.cpp_info.names["cmake_find_package_multi"] = "z3"
```
Comment (jgsogo) [inline recipes/z3/all/conanfile.py:86]: Maybe need:

```suggestion
        self.cpp_info.names["cmake_find_package_multi"] = "z3"
        self.cpp_info.filenames["cmake_find_package"] = "Z3"
        self.cpp_info..filenames["cmake_find_package_multi"] = "Z3"
```

File: recipes/z3/all/test_package/CMakeLists.txt [hunk 6]
```diff
@@ -0,0 +1,8 @@
+cmake_minimum_required(VERSION 3.1)
+project(test_package)
+
+include(${CMAKE_BINARY_DIR}/conanbuildinfo.cmake)
+conan_basic_setup()
```
Comment (madebr) [inline recipes/z3/all/test_package/CMakeLists.txt:5]:
```suggestion
conan_basic_setup(TARGETS)

find_package(Z3 REQUIRED)
```

File: recipes/z3/all/test_package/CMakeLists.txt [hunk 7]
```diff
@@ -0,0 +1,8 @@
...
+
+add_executable(${PROJECT_NAME} test_package.cpp)
+target_link_libraries(${PROJECT_NAME} ${CONAN_LIBS})
```
Comment (madebr) [inline recipes/z3/all/test_package/CMakeLists.txt:8]:
```suggestion
target_link_libraries(${PROJECT_NAME} PRIVATE z3::libz3)
```

File: recipes/z3/all/test_package/conanfile.py [hunk 8]
```diff
@@ -0,0 +1,17 @@
+from conans import ConanFile, CMake, tools
+import os
+
+
+class TestPackageConan(ConanFile):
+    settings = "os", "compiler", "build_type", "arch"
+    generators = "cmake"
```
Comment (madebr) [inline recipes/z3/all/test_package/conanfile.py:7]:
```suggestion
    generators = "cmake", "cmake_find_package"
```

File: recipes/z3/all/conanfile.py (hunk 5 above)
Comment (madebr) [inline recipes/z3/all/conanfile.py:86]:
```suggestion
        self.cpp_info.names["cmake_find_package"] = "z3"
        self.cpp_info.names["cmake_find_package_multi"] = "z3"
        self.cpp_info.filenames["cmake_find_package"] = "Z3"
        self.cpp_info.filenames["cmake_find_package_multi"] = "Z3"
```

Comment (madebr) [reply recipes/z3/all/conanfile.py:86]: Applied, minus the typo.

File: recipes/z3/all/conanfile.py (hunk 1 above)
Comment (jgsogo) [reply recipes/z3/all/conanfile.py:98]: There is no way to create more than one `FindXXXX.cmake` file. If this is something needed/desirable, we need a feature in Conan.

**Review by jgsogo — APPROVED** [review]

**Review by danimtb — APPROVED** [review]

**Pull request merged**